Source/AsyncGenerator/Internal/EnvironmentHelper.cs

```diff
@@ -102,5 +99,43 @@ public static string GetNetCoreMsBuildPath()
 			var msBuildPath = Path.Combine(dotnetFolder, "sdk", netCoreVersion, "MSBuild.dll");
 			return File.Exists(msBuildPath) ? msBuildPath : null;
 		}
+
+		public static string GetMonoMsBuildPath(Action<string> monoDirectoryAction = null)
+		{
+			// Get the sdk path by using the Mono runtime assembly location
+			// Default locations:
+			// Windows -> file:///C:/Program Files (x86)/Mono/lib/mono/4.5/mscorlib.dll
+			// Linux -> file:///usr/lib/mono/4.5/mscorlib.dll
+			// OSX -> file:///Library/Frameworks/Mono.framework/Versions/5.2.0/lib/mono/4.5/mscorlib.dll
+			// MSBuild.dll is then located:
+			// Windows -> C:/Program Files (x86)/Mono/lib/mono/msbuild/15.0/bin/MSBuild.dll
+			// Linux -> /usr/lib/mono/msbuild/15.0/bin/MSBuild.dll
+			// OSX -> /Library/Frameworks/Mono.framework/Versions/5.2.0/lib/mono/msbuild/15.0/bin/MSBuild.dll
+
+			var assembly = typeof(System.Runtime.GCSettings).Assembly;
+			// Remove file:// on Unix or file:/// on Windows
+			var assemblyDirectory = Path.GetDirectoryName(assembly.CodeBase.Substring(IsWindows ? 8 : 7));
```

Comment:
var assemblyDirectory = new Uri(assembly.CodeBase).LocalPath;

Reply:
https://stackoverflow.com/questions/4107625/how-can-i-convert-assembly-codebase-into-a-filesystem-path-in-c

Reply:
Also, please check this: https://blogs.msdn.microsoft.com/suzcook/2003/06/26/assembly-codebase-vs-assembly-location/

Reply:
:+1: